File: verification/probes/sdt_onnx_replay.mjs
// Run from the isolated document-worker checkout; capture tensors first.
import fs from 'node:fs';
import { deserialize } from 'node:v8';
import { createRequire } from 'node:module';
import { createHash } from 'node:crypto';
import { performance } from 'node:perf_hooks';
const ort = createRequire(import.meta.url)(process.env.SDT_ORT_PATH);
const calls = deserialize(fs.readFileSync(process.argv[2]));
const backend = process.env.SDT_BACKEND;
const models = ['clusterer/model.onnx','clusterer/repair.onnx','classifier/model.onnx'];
const sessions = new Map();
for (const model of models) {
 const bytes = fs.readFileSync(`build/block-seg/${model}`);
 const hash = createHash('sha256').update(bytes).digest('hex');
 sessions.set(hash, await ort.InferenceSession.create(bytes, {executionProviders:[backend],intraOpNumThreads:1,interOpNumThreads:1,graphOptimizationLevel:'all',...(process.env.SDT_PROFILE ? {enableProfiling:true,profileFilePrefix:`${process.env.SDT_PROFILE}/${hash}`} : {})}));
}
const rows = [];
for (let i=0;i<6;i++) {
 let maxAbsError=0, sumMs=0; const byModel={};
 for (const call of calls) {
  const feeds=Object.fromEntries(Object.entries(call.feeds).map(([k,t])=>[k,new ort.Tensor(t.type,t.data,t.dims)]));
  const start=performance.now();
  const out=await sessions.get(call.hash).run(feeds,call.outputNames);
  const ms=performance.now()-start; sumMs+=ms; byModel[call.hash]=(byModel[call.hash]||0)+ms;
  for (const [key,tensor] of Object.entries(out)) {
   const expected=call.outputs[key].data;
   if (tensor.data.length!==expected.length) throw new Error('Output shape changed');
   for(let j=0;j<expected.length;j++) {
    const error=Math.abs(Number(tensor.data[j])-Number(expected[j]));
    if (!Number.isFinite(error)) throw new Error('Nonfinite output difference');
    maxAbsError=Math.max(maxAbsError,error);
   }
   tensor.dispose();
  }
 }
 rows.push({iteration:i,warmup:i===0,sumMs,byModel,maxAbsError});
}
for (const session of sessions.values()) { if(process.env.SDT_PROFILE)session.endProfiling(); await session.release(); }
fs.writeFileSync(process.argv[3],JSON.stringify({backend,calls:calls.length,rows},null,2));
console.log(JSON.stringify(rows));

File: verification/probes/sdt_onnx_replay.py
"""Replay real SDT tensors with explicit CPU or CUDA; no PDF parser is loaded."""
import argparse
import hashlib
import json
from pathlib import Path
import time

import numpy as np
import onnxruntime as ort

parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument('capture', type=Path)
parser.add_argument('output', type=Path)
parser.add_argument('--backend', choices=['cpu', 'cuda'], required=True)
parser.add_argument('--profile-dir', type=Path)
args = parser.parse_args()
provider = 'CUDAExecutionProvider' if args.backend == 'cuda' else 'CPUExecutionProvider'
if provider not in ort.get_available_providers():
    raise RuntimeError(f'{provider} unavailable')
options = ort.SessionOptions()
options.intra_op_num_threads = 1
options.inter_op_num_threads = 1
options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
sessions = {}
for model in ['clusterer/model.onnx', 'clusterer/repair.onnx', 'classifier/model.onnx']:
    path = Path('build/block-seg') / model
    digest = hashlib.sha256(path.read_bytes()).hexdigest()
    if args.profile_dir:
        args.profile_dir.mkdir(parents=True, exist_ok=True)
        options.enable_profiling = True
        options.profile_file_prefix = str(args.profile_dir / digest)
    session = ort.InferenceSession(str(path), sess_options=options, providers=[provider])
    if provider not in session.get_providers():
        raise RuntimeError(f'{model}: requested provider was not loaded')
    sessions[digest] = session
calls = json.loads(args.capture.read_text())
prepared = []
for call in calls:
    feeds = {key: np.array(value['data'], dtype=value['type']).reshape(value['dims'])
             for key, value in call['feeds'].items()}
    prepared.append((call, feeds))
rows = []
for iteration in range(6):
    total_ms = 0
    by_model = {}
    max_abs_error = 0
    for call, feeds in prepared:
        session = sessions[call['hash']]
        names = call.get('outputNames') or [out.name for out in session.get_outputs()]
        start = time.perf_counter()
        outputs = session.run(names, feeds)
        elapsed = (time.perf_counter() - start) * 1000
        total_ms += elapsed
        by_model[call['hash']] = by_model.get(call['hash'], 0) + elapsed
        for name, output in zip(names, outputs, strict=True):
            expected = np.array(call['outputs'][name]['data']).reshape(output.shape)
            if output.size:
                error = float(np.max(np.abs(output.astype(float) - expected)))
                if not np.isfinite(error):
                    raise RuntimeError('Nonfinite output difference')
                max_abs_error = max(max_abs_error, error)
    rows.append({'iteration': iteration, 'warmup': iteration == 0, 'sumMs': total_ms,
                 'byModel': by_model, 'maxAbsError': max_abs_error})
profiles = [session.end_profiling() for session in sessions.values()] if args.profile_dir else []
args.output.write_text(json.dumps({'backend': args.backend, 'ortVersion': ort.__version__,
                                  'calls': len(calls), 'rows': rows, 'profiles': profiles}, indent=2))
print(json.dumps(rows), flush=True)
